### PR TITLE
:running: enable e2e tests on travis

### PR DESCRIPTION
**Pull request by mengqiy**
This PR enables e2e test for v1 project.
Some code under `/test/e2e` was accidentally deleted in an earlier refactoring. I introduce it back with necessary changes.

### CODE REVIEW

**Comment by mengqiy** [comment]
PTAL
It is working now.

File: .travis.yml [hunk 1]
```diff
@@ -18,13 +18,19 @@ git:
 
 go_import_path: sigs.k8s.io/kubebuilder
 
+services:
+- docker
+
 before_install:
 - go get -u github.com/golang/dep/cmd/dep
 - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then curl -sL https://github.com/git-lfs/git-lfs/releases/download/v2.7.2/git-lfs-darwin-amd64-v2.7.2.tar.gz | tar -xz git-lfs; fi
+- if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then GO111MODULE=on go get sigs.k8s.io/kind && kind create cluster --config test/kind-config.yaml && export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"; fi
```
Comment (droot) [inline .travis.yml]: use `kind@0.3.0` and `kustomize@SHA1`

File: .travis.yml [hunk 2]
```diff
@@ -18,13 +18,19 @@ git:
 
 go_import_path: sigs.k8s.io/kubebuilder
 
+services:
+- docker
```
Comment (droot) [inline .travis.yml]: Check if we can enable caching for docker images ? will make it run faster.

File: test/e2e/e2e.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+				"GO111MODULE=off",
+				"IMG=controller:v0.1.0",
+				"KUBECONFIG=" + strings.TrimSpace(string(out)),
+			})
+
+		prepare(c.workDir)
+		defer cleanupv1(kubebuilderTest, c.workDir, c.controllerImageName)
+
+		// prepare v1 vendor
+		log.Printf("running: %s", "tar -zxf ../../../testdata/vendor.v1.tgz")
```
Comment (droot) [inline test/e2e/e2e.go]: By ?

File: test/internal/e2e/kubebuildertest.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	cmd.Dir = kt.Dir
+	cmd.Env = append(os.Environ(), kt.Env...)
+	command := strings.Join(cmd.Args, " ")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return output, fmt.Errorf("%s failed with error: %s", command, string(output))
+	}
+	log.Printf("%s finished successfully", command)
+	return output, nil
+}
```
Comment (droot) [inline test/internal/e2e/kubebuildertest.go]: Can you pl. get rid of unused functions.

File: test/e2e/e2e.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+		err = kubebuilderTest.CreateAPI(coreControllerOptions)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("building image")
+		log.Printf("running: %s", "make docker-build")
+		makeDockerBuildOptions := []string{"docker-build"}
+		err = kubebuilderTest.Make(makeDockerBuildOptions)
+		Expect(err).NotTo(HaveOccurred())
+
+		log.Printf("running: %s", "kind load docker-image controller:v0.1.0")
```
Comment (droot) [inline test/e2e/e2e.go]: By ?

File: test/e2e/framework/ginkgowrapper/wrapper.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+			scanner.Scan() // these come in pairs
+		} else {
+			prunedStack = append(prunedStack, scanner.Text())
+			scanner.Scan() // these come in pairs
+			prunedStack = append(prunedStack, scanner.Text())
+		}
+	}
+
+	return strings.Join(prunedStack, "\n")
+}
```
Comment (droot) [inline test/e2e/framework/ginkgowrapper/wrapper.go]: Pl. check if these functions are still used ?

File: .travis.yml (hunk 2 above)
Comment (mengqiy) [reply .travis.yml]: Travis document doesn't recommend caching docker images: https://docs.travis-ci.com/user/caching/#things-not-to-cache

File: test/e2e/framework/util.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"math/rand"
+	"strings"
+	"time"
+)
+
+// RandomSuffix provides a random string to append to certain base name.
+func RandomSuffix() string {
+	source := []rune("abcdefghijklmnopqrstuvwxyz")
```
Comment (DirectXMan12) [inline test/e2e/framework/util.go]: this... is insufficient.  We're seeding on the current time, which means parallel tests might get the same seed.

Use kube to our advantage -- generate a random hash, try to grab a "lock" by creating a cluster-scoped object, then if we fail try again with something new.  Or use generatename on a cluster-scoped object -- those are guaranteed not to conflict.

File: test/e2e/framework/util.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	res := make([]rune, 4)
+	for i := range res {
+		res[i] = source[r.Intn(len(source))]
+	}
+	return string(res)
+}
+
+// ParseCmdOutput converts given command output string into individual objects
+// according to line breakers, and ignores the empty elements in it.
+func ParseCmdOutput(output string) []string {
```
Comment (DirectXMan12) [inline test/e2e/framework/util.go]: this needs a more descriptive name

File: test/e2e/kubebuildertest.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+)
+
```
Comment (DirectXMan12) [inline test/e2e/kubebuildertest.go]: godoc

File: test/e2e/kubebuildertest.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	// environment variables in k=v format.
+	Env []string
+}
+
+func NewKubebuilderTest(dir, binDir string, env []string) *KubebuilderTest {
+	kt := KubebuilderTest{
+		Dir: dir,
+		Env: env,
+	}
+	os.Setenv("TEST_ASSET_KUBECTL", strings.Join([]string{binDir, "kubectl"}, "/"))
```
Comment (DirectXMan12) [inline test/e2e/kubebuildertest.go]: why are we setting these instead of respecting the environment?

**Comment by mengqiy** [comment]
PTAL

File: scripts/install-dependencies.sh [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+go get sigs.k8s.io/kind@v0.3.0
+# You can use --image flag to specify the cluster version you want, e.g --image=kindest/node:v1.13.6, the supported version are listed at https://hub.docker.com/r/kindest/node/tags
+kind create cluster --config test/kind-config.yaml --image=kindest/node:v1.14.1
+
+# TODO(mengqiy): use a released version after v2.1.0 is out.
+go get sigs.k8s.io/kustomize@f9c631e9eec7a2d6e46eb9e1bf5122f68b97d12d
```
[the remark this answers is not in the document]
Comment (mengqiy) [reply scripts/install-dependencies.sh]: Done

**Comment by mengqiy** [comment]
Squashed commits.
PTAL

File: go.mod [hunk 12]
```diff
@@ -3,18 +3,32 @@ module sigs.k8s.io/kubebuilder
...
 	golang.org/x/text v0.3.2 // indirect
 	golang.org/x/tools v0.0.0-20190506145303-2d16b83fe98c
 	gopkg.in/yaml.v2 v2.2.2
+	k8s.io/api v0.0.0-20190409021203-6e4e0e4f393b // indirect
+	k8s.io/apimachinery v0.0.0-20190404173353-6a84e37a896d // indirect
+	k8s.io/client-go v11.0.0+incompatible // indirect
+	k8s.io/klog v0.3.0
+	k8s.io/kube-openapi v0.0.0-20190510232812-a01b7d5d6c22 // indirect
+	sigs.k8s.io/kind v0.1.0 // indirect
+	sigs.k8s.io/kustomize v0.0.0-20190516203229-f9c631e9eec7 // indirect
```
Comment (DirectXMan12) [inline go.mod]: can you run `go mod tidy`?

File: scripts/install-dependencies.sh [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+go get sigs.k8s.io/kind@v0.3.0
+# TODO(mengqiy): use a released version after v2.1.0 is out.
+go get sigs.k8s.io/kustomize@f9c631e9eec7a2d6e46eb9e1bf5122f68b97d12d
+
+# You can use --image flag to specify the cluster version you want, e.g --image=kindest/node:v1.13.6, the supported version are listed at https://hub.docker.com/r/kindest/node/tags
+kind create cluster --config test/kind-config.yaml --image=kindest/node:v1.14.1
```
Comment (DirectXMan12) [inline scripts/install-dependencies.sh]: you should name this script something else if it's creating a cluster -- `install-dependencies` implies it's just installing things.

File: test/e2e/e2e_test.go [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// Run e2e tests using the Ginkgo runner.
+func TestE2E(t *testing.T) {
+	RegisterFailHandler(Fail)
+	fmt.Fprintf(GinkgoWriter, "Starting kubebuilder suite\n")
```
Comment (DirectXMan12) [inline test/e2e/e2e_test.go]: ??

Comment (DirectXMan12) [reply test/e2e/e2e_test.go]: still an issue

File: test/e2e/e2ev1.go [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+			Expect(err).NotTo(HaveOccurred())
+			kbTest = &kubebuilderTest{
+				Dir: c.workDir,
+				Env: []string{"GO111MODULE=off"},
+			}
+			prepare(c.workDir)
+		})
+
+		AfterEach(func() {
+			cleanupv1(kbTest, c.workDir, c.controllerImageName)
```
Comment (DirectXMan12) [inline test/e2e/e2ev1.go]: why isn't this just inline here?

File: test/e2e/e2ev1.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+			err = kbTest.Make("deploy")
+			Expect(err).Should(Succeed())
+
+			By("validate the controller-manager pod running as expected")
+			verifyControllerUp := func() error {
+				// Get pod name
+				podOutput, err := kbTest.RunKubectlCommand(
+					"get", "pods",
+					"-l", "control-plane=controller-manager",
+					"-n", fmt.Sprintf("e2e-%s-system", testSuffix),
```
Comment (DirectXMan12) [inline test/e2e/e2ev1.go]: move this to a variable or a method, since you do it a few times

File: test/e2e/util.go [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+	. "github.com/onsi/gomega"
+)
+
+func init() {
+	rand.Seed(time.Now().UTC().UnixNano())
+}
+
+// randomSuffix provides a random string to append to certain base name.
+func randomSuffix() string {
+	source := []rune("abcdefghijklmnopqrstuvwxyz")
```
Comment (DirectXMan12) [inline test/e2e/util.go]: same comments from before re: random suffix unless I missed something else

Comment (mengqiy) [reply test/e2e/util.go]: Before: Seed with current time in each `randomSuffix` function call. 2 parallel tests may end up using the same seed.
Now: Seed with current time in the `init` function. It should be fine now. Correct me if I'm wrong :)

File: go.mod (hunk 12 above)
Comment (mengqiy) [reply go.mod]: Much better now :)

File: test/e2e/e2e_test.go (hunk 14 above)
Comment (mengqiy) [reply test/e2e/e2e_test.go]: I'd like to discuss this with you in person :)

File: test/e2e/util.go (hunk 17 above)
Comment (mengqiy) [reply test/e2e/util.go]: Fixed

**Comment by mengqiy** [comment]
PTAL

**Comment by mengqiy** [comment]
Hmm... It seems I can only use lower case letters. Digits are disallowed. I will update the PR shortly.

**Comment by mengqiy** [comment]
Tests are passing. PTAL

**Comment by mengqiy** [comment]
@droot @DirectXMan12 Can someone please tag the label?

**Pull request merged**